### ISSUE TITLE
install_mp_handler() leaking semaphores?

### ISSUE THREAD

**Issue by plamut** [jruere/multiprocessing-logging#62]
In production we execute various periodic background jobs using the [AP Scheduler](https://pypi.org/project/APScheduler/) package. The app runs in a docker container. We add periodic jobs to the scheduler at application startup by calling the scheduler's `add_job()` method:

```py
self.scheduler.add_job(
    _execute,
    args=job_args,
    trigger="interval",
    max_instances=max_instances,
    # trigger arguments below
    seconds=interval.total_seconds(),
    start_date=start_time,
    name=name,
)
```

The `_execute()` function that we submit is our own wrapper around the actual callable that contains the job's core logic. Before invoking the latter, we initialize multiprocessing logging:
```python
try:
  # The following literally just calls  multiprocessing_logging.install_mp_handler(), and that's it
   context.setup_multiprocess_logging()
except OSError as exc:
    logger.exception(exc)

    curr_proc = psutil.Process(os.getpid())
    open_files = curr_proc.num_fds()
    max_open_files = curr_proc.rlimit(psutil.RLIMIT_NOFILE)
    logger.error(
        (
            "Setting up multiprocess logging failed: %s"
            "Process using %d file handle(s), RLIMIT_NOFILE: %s"
        ),
        exc, open_files, max_open_files
    )
    raise
```

However, every now and then we observer that setting up multiprocessing logging fails. Unpredictable, when, but it eventually always occurs, and we need to restart the app:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/src/oracle/dex_ohlcv/scheduler/scheduler.py", line 189, in _execute
    context.setup_multiprocess_logging()
  File "/usr/src/oracle/dex_ohlcv/processcontext.py", line 314, in setup_multiprocess_logging
    install_mp_handler()
  File "/usr/local/lib/python3.10/site-packages/multiprocessing_logging.py", line 27, in install_mp_handler
    handler = MultiProcessingHandler("mp-handler-{0}".format(i), sub_handler=orig_handler)
  File "/usr/local/lib/python3.10/site-packages/multiprocessing_logging.py", line 60, in __init__
    self.queue = multiprocessing.Queue(-1)
  File "/usr/local/lib/python3.10/multiprocessing/context.py", line 103, in Queue
    return Queue(maxsize, ctx=self.get_context())
  File "/usr/local/lib/python3.10/multiprocessing/queues.py", line 48, in __init__
    self._wlock = ctx.Lock()
  File "/usr/local/lib/python3.10/multiprocessing/context.py", line 68, in Lock
    return Lock(ctx=self.get_context())
  File "/usr/local/lib/python3.10/multiprocessing/synchronize.py", line 162, in __init__
    SemLock.__init__(self, SEMAPHORE, 1, 1, ctx=ctx)
  File "/usr/local/lib/python3.10/multiprocessing/synchronize.py", line 57, in __init__
    sl = self._semlock = _multiprocessing.SemLock(
OSError: [Errno 28] No space left on device
```

There is plenty of free storage space on the server, thus _"No space on device"_ likely means that we run out of semaphore locks, we suspect a resource leak somewhere.

In the exception handler we also log the number of open file descriptors, and while the number is high, it's still way below the system limit:
```
Setting up multiprocess logging failed: [Errno 28] No space left on deviceProcess using 1265 file handle(s), RLIMIT_NOFILE: (1048576, 1048576)
```

**Questions:**
- Do we use `install_mp_handler()` correctly, i.e. every time the AP Scheduler starts executing a job in a new child process?
- Are you aware of any potential semaphore leaks by using `install_mp_handler()` this way? Are we even exploring in the right direction?
- Does logging the number of currently open file descriptors even make sense? Do those include the any semaphore locks?

---

TL; DR - the crashes always happen at the same place, i.e. when calling `install_mp_handler()`. We suspect a resource leak, but not sure if it's caused by `install_mp_handler()` internals, or by the way we use it.

We appreciate any tips or insights, and we can of course provide any additional information needed.
If this is definitely **not** an issue with `install_mp_handler()`, that's also OK, we can at least focus our search elsewhere.

**Comment by jruere** [comment]
I've never seen this problem before! :D


>    Do we use install_mp_handler() correctly, i.e. every time the AP Scheduler starts executing a job in a new child process?

It should only be called once during the execution of the main thread. I don't know this framework so I don't know what's happening.

In case you want this functionality only for some scheduled jobs, you could install the handler and then remove it when the job is done.

>    Are you aware of any potential semaphore leaks by using install_mp_handler() this way? Are we even exploring in the right direction?

Semaphores are not an OS resource, AFAIK, but I could be wrong. I suspect you are running out of shared memory. Could you check how much shared memory is in use when you see this error and shortly after the AP Scheduler is started?

>    Does logging the number of currently open file descriptors even make sense? Do those include the any semaphore locks?

File descriptors must be used as part of IPC but the main process should use very few total when used correctly.


I suspect you are installing the handler multiple times during the lifetime of the main process and this stacks new handlers in the logging system (you can check this too). Each handler has a multiprocessing queue and a thread. The mp queue will use some IPC; which probably uses shared memory.

**Comment by plamut** [comment]
First of all, thanks a lot for the reply, and for several useful hints on what to check!

> I've never seen this problem before! :D

Indeed, found nothing on the internet, thus the culprit is most likely hiding somewhere in our own code.
Will add logging of more system resource stats, but of course it might take some time until we get the next reproduction.

Thanks again for now!

**Comment by plamut** [comment]
I found out that there exists a  `_POSIX_SEM_NSEMS_MAX` limit, which is 255 (?) by default.

---

**Edit:**
Or maybe unlimited:
```py
>>> import os
>>> os.sysconf("SC_SEM_NSEMS_MAX")
-1
```

---

And even if that limit is higher, the semaphores indeed consume shared memory, and creating too many of them eventually exhausts that resource:
https://man7.org/linux/man-pages/man7/sem_overview.7.html
```
Unnamed semaphores (memory-based semaphores)
              An unnamed semaphore does not have a name.  Instead the
              semaphore is placed in a region of memory that is shared
              between multiple threads (a thread-shared semaphore) or
              processes (a process-shared semaphore). 
             ...
              A process-shared semaphore must be placed in a shared
              memory region (e.g., a System V shared memory segment
              created using [shmget(2)](https://man7.org/linux/man-pages/man2/shmget.2.html),
              or a POSIX shared memory 
```

In any case, it's most likely something in our system that hits one of the resource limits, as opposed to the library itself. I will thus close the issue.

I _might_ follow up with the findings if/when we get to the bottom of this, just in case someone else finds them useful in the future.

**Comment by plamut** [comment]
Just posting a follow-up in case it helps anyone in the future - the most likely issue was that we invoked `install_mp_handler()` in every child process, i.e. from withing the job that started executing. No wonder we eventually ran out of system resources! :man_facepalming:

Moving the `install_mp_handler()` call to before the `scheduler` was instantiated (which internally creates a process pool) seems to have helped. In less than a week we observed the error on all production instances, except on the one the change was deployed on. This made us believe that an incorrectly placed `install_mp_handler()` call was indeed the culprit.

In the end we did not bother figuring out which particular resource we ran out of, let this be left as an exercise for the future reader. :upside_down_face:

**Comment by jruere** [comment]
Thanks for the update!
